File: core/src/main/java/fr/sncf/osrd/api/pathfinding/response/PathfindingResult.java
package fr.sncf.osrd.api.pathfinding.response;

import com.squareup.moshi.Json;
import com.squareup.moshi.JsonAdapter;
import com.squareup.moshi.Moshi;
import fr.sncf.osrd.railjson.schema.common.ID;
import fr.sncf.osrd.railjson.schema.geom.RJSLineString;
import fr.sncf.osrd.railjson.schema.infra.RJSRoutePath;
import fr.sncf.osrd.reporting.warnings.Warning;
import java.util.*;

public class PathfindingResult {
    public static final JsonAdapter<PathfindingResult> adapterResult = new Moshi
            .Builder()
            .add(ID.Adapter.FACTORY)
            .build()
            .adapter(PathfindingResult.class)
            .failOnUnknown();
    @Json(name = "route_paths")
    public List<RJSRoutePath> routePaths = new ArrayList<>();
    @Json(name = "path_waypoints")
    public List<PathWaypointResult> pathWaypoints = new ArrayList<>();

    public final double length;

    public RJSLineString geographic = null;
    public RJSLineString schematic = null;

    public List<SlopeChartPointResult> slopes = new ArrayList<>();
    public List<CurveChartPointResult> curves = new ArrayList<>();

    public List<Warning> warnings = null;

    public PathfindingResult(double length) {
        this.length = length;
    }
}

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/response/PathfindingResult.kt
package fr.sncf.osrd.api.pathfinding.response

import com.squareup.moshi.Json
import com.squareup.moshi.JsonAdapter
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import fr.sncf.osrd.railjson.schema.common.ID
import fr.sncf.osrd.railjson.schema.geom.RJSLineString
import fr.sncf.osrd.railjson.schema.infra.RJSRoutePath
import fr.sncf.osrd.reporting.exceptions.OSRDError
import fr.sncf.osrd.reporting.warnings.Warning

class PathfindingResult(@JvmField val length: Double) {
    @JvmField
    @Json(name = "route_paths")
    var routePaths: List<RJSRoutePath> = ArrayList()

    @Json(name = "path_waypoints")
    var pathWaypoints: List<PathWaypointResult> = ArrayList()

    var geographic: RJSLineString? = null

    var schematic: RJSLineString? = null

    var slopes: List<SlopeChartPointResult> = ArrayList()

    var curves: List<CurveChartPointResult> = ArrayList()

    var warnings: List<Warning>? = null

    companion object {
        val adapterResult: JsonAdapter<PathfindingResult> = Moshi.Builder()
            .add(KotlinJsonAdapterFactory())
            .add(ID.Adapter.FACTORY)
            .build()
            .adapter(PathfindingResult::class.java)
            .failOnUnknown()
    }
}

enum class ResponseState {
    SUCCESS, ERROR
}

class PathfindingResponse(val responseState: ResponseState, val errorMessage: String?, val pathfindingResult: PathfindingResult?) {
    companion object {
        val adapterResult: JsonAdapter<PathfindingResponse> = Moshi.Builder()
            .add(KotlinJsonAdapterFactory())
            .add(ID.Adapter.FACTORY)
            .build()
            .adapter(PathfindingResponse::class.java)
            .failOnUnknown()
    }
}